File: 198-house-robber/198-house-robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        unordered_map<int, int> memo;
        memo[n-1] = nums[n-1];
        for (int i = n-2; i >= 0; i--) {
            memo[i] = max(memo[i+2]+nums[i], memo[i+1]);
        }
        
        return memo[0];
    }  
};

File: 198-house-robber/198-house-robber.java
class Solution {
    int[] nums;
    Integer[] memo;
    public int rob(int[] nums) {
        this.nums = nums;
        this.memo = new Integer[nums.length];
        return dfs(nums.length-1);
    }
    
    private int dfs(int i) {
        if (i < 0) {
            return 0;
        }
        if (memo[i] == null) {
            memo[i] = Math.max(dfs(i-2)+nums[i], dfs(i-1));
        }
        return memo[i];
    }
}